src/array_unique.rs

```diff
@@ -0,0 +1,49 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2022
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use pyo3::prelude::*;
+use pyo3::wrap_pyfunction;
+
+use hashbrown::HashMap;
+use numpy::PyReadonlyArray2;
+
+#[pyfunction]
+#[pyo3(text_signature = "(array)")]
+pub fn unique(array: PyReadonlyArray2<u16>) -> PyResult<(Vec<usize>, Vec<usize>)> {
+    let array = array.as_array();
+    let shape = array.shape();
+    let mut table = HashMap::<Vec<u16>, usize>::new();
```

Comment:
You can preallocate this since table will be of size `shape[0]`:

```suggestion
    let mut table = HashMap::<Vec<u16>, usize>::with_capacity(shape[0]);
```

Reply:
Actually never mind, that will potentially over allocate. So you probably don't want to do this

Reply:
Presumably `HashMap` amortises the allocation cost to avoid a quadratic complexity?  I'd expect it to only re-allocate `O(log(shape[0]))` times at most.

Reply:
Thanks. `with_capacity(shape[0])` is faster than `new()` significantly.

`with_capacity(shape[0])`
```
8, 1000 (1st pass): 0.159495551 sec
8, 1000 (2nd pass): 0.008254019000000001 sec

16, 1000 (1st pass): 0.26337141 sec
16, 1000 (2nd pass): 0.11570970999999997 sec

24, 1000 (1st pass): 0.1347217279999997 sec
24, 1000 (2nd pass): 0.10911709100000033 sec

32, 1000 (1st pass): 0.12337133300000014 sec
32, 1000 (2nd pass): 0.09941815200000015 sec

40, 1000 (1st pass): 0.13044818299999994 sec
40, 1000 (2nd pass): 0.12741578799999997 sec
```

`new()`
```
8, 1000 (1st pass): 0.12296806400000004 sec
8, 1000 (2nd pass): 0.008027125000000024 sec

16, 1000 (1st pass): 0.34793314100000017 sec
16, 1000 (2nd pass): 0.22817380199999993 sec

24, 1000 (1st pass): 0.17599335599999977 sec
24, 1000 (2nd pass): 0.165346494 sec

32, 1000 (1st pass): 0.2335444459999998 sec
32, 1000 (2nd pass): 0.1701327779999997 sec

40, 1000 (1st pass): 0.23492698999999995 sec
40, 1000 (2nd pass): 0.20621412699999997 sec
```